test/e2e_node/image_list.go

```diff
@@ -151,23 +155,59 @@ func PrePullAllImages() error {
 	}
 	images := framework.ImageWhiteList.List()
 	klog.V(4).Infof("Pre-pulling images with %s %+v", puller.Name(), images)
-	for _, image := range images {
-		var (
-			err    error
-			output []byte
-		)
-		for i := 0; i < maxImagePullRetries; i++ {
-			if i > 0 {
-				time.Sleep(imagePullRetryDelay)
-			}
-			if output, err = puller.Pull(image); err == nil {
-				break
+
+	imageCh := make(chan int, len(images))
+	for i := range imageCh {
+		imageCh <- i
+	}
+	close(imageCh)
+
+	pullErrs := make([]error, len(images))
+	pullProgressStoped := int32(0)
+
+	parallelImagePullCount := maxParallelImagePullCount
+	if len(images) < parallelImagePullCount {
+		parallelImagePullCount = len(images)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(parallelImagePullCount)
+	for i := 0; i < parallelImagePullCount; i++ {
+		go func() {
+			defer wg.Done()
+
+			for i := range imageCh {
+				var (
+					pullErr error
+					output  []byte
+				)
+				for retryCount := 0; retryCount < maxImagePullRetries; retryCount++ {
+					if atomic.LoadInt32(&pullProgressStoped) == 1 {
```

Comment:
I'm open to using `atomic` here - however, it seems like we're trying to communicate via shared memory, and my understanding is that Golang tries to prefer channels instead of shared memory.

Thoughts on having a `quit` channel, and then whenever we want to stop, we can send a signal to the `quit` channel? We could use a `select` statement with a fall through to reattempting the pull as the default case.

Let me know if a code sample could be helpful :)

Reply:
Use channels instead of shared memory is recommended in Golang,  and yes, here we can use channel to solve this problem.

But If we use channel there are other points that must be considered, we can't close the **quite** channel to notify other goroutines to stop(because other goroutines maybe close the **quite** channel as well, double-close will panic), and we should do something to fix this(and here, we go back with two-option: shared memory or channels :)).

If we send a signal to the **quit** channel, we must send at-least **parallelImagePullCount** values, to ensure all goroutine will receive one value from the channel. I think this is a little complicated to solve the problem.

Finally, we can use **context** and cancel to simplify the solution, the **cannel** function support multi called. Do you think this will be better?

Reply:
Can you share more around where you're reading about contexts being able to be cancelled more than once? https://www.sohamkamani.com/golang/2018-06-17-golang-using-context-cancellation/#gotchas-and-caveats seems to suggest otherwise.

Another option is that we could set up a channel in the main goroutine on which the worker goroutines can send errors back to the main goroutine. The first time that channel receives an error, it would close the `quit` channel. In that way, only a single thread can try and close the `quit` channel, and we don't need to worry about race conditions. It also has the side benefit of returning the errors to the main goroutine.

Reply:
Got this from the comment in [golang project source code](https://github.com/golang/go/blob/master/src/context/context.go#L222).

Reply:
Ah, that certainly seems pretty convincing :)

Mind just testing to verify that calling `cancel` multiple times works ok? If yes, than using a context sounds perfect.

Reply:
Tested with 100 goroutines(each invokes **cancel** 100 times), works as expected. 

Refactored atomic to context, PTAL :)